runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java

```diff
@@ -0,0 +1,197 @@
+  @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+  public void expandMultipleTest() {
+    PCollection<Integer> pcol =
+        testPipeline
+            .apply(Create.of(1, 2, 3))
+            .apply("add one", External.of(TEST_URN_SIMPLE, new byte[] {}, localExpansionAddr))
+            .apply(
+                "filter <=3",
+                External.of(TEST_URN_LE, "3".getBytes(StandardCharsets.UTF_8), localExpansionAddr));
+
+    PAssert.that(pcol).containsInAnyOrder(2, 3);
+    testPipeline.run();
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+  public void expandMultiOutputTest() {
+    PCollectionTuple pTuple =
+        testPipeline
+            .apply(Create.of(1, 2, 3, 4, 5, 6))
+            .apply(
+                External.of(TEST_URN_MULTI, new byte[] {}, localExpansionAddr).withMultiOutputs());
+
+    PAssert.that(pTuple.get(new TupleTag<Integer>("aTag") {})).containsInAnyOrder(2, 4, 6);
+    PAssert.that(pTuple.get(new TupleTag<Integer>("bTag") {})).containsInAnyOrder(1, 3, 5);
+    testPipeline.run();
+  }
+
+  private Process runCommandline(String command) {
+    ProcessBuilder builder = new ProcessBuilder("sh", "-c", command);
+    try {
+      return builder.start();
+    } catch (IOException e) {
+      throw new AssertionError("process launch failed.");
+    }
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+  public void expandPythonTest() {
+    String target = String.format("localhost:%s", expansionPort);
+    Process p = runCommandline(String.format("%s -p %s", pythonServerCommand, expansionPort));
+    try {
+      ManagedChannel channel = ManagedChannelBuilder.forTarget(target).build();
+      ConnectivityState state = channel.getState(true);
+      for (int retry = 0; retry < 30 && state != ConnectivityState.READY; retry++) {
+        Thread.sleep(500);
+        state = channel.getState(true);
+      }
+      channel.shutdownNow();
+
+      PCollection<String> pCol =
+          testPipeline
+              .apply(Create.of("1", "2", "2", "3", "3", "3"))
+              .apply(
+                  "toBytes",
+                  MapElements.into(new TypeDescriptor<byte[]>() {}).via(String::getBytes))
+              .apply(External.<byte[]>of("count_per_element_bytes", new byte[] {}, target))
+              .apply("toString", MapElements.into(TypeDescriptors.strings()).via(String::new));
+
+      PAssert.that(pCol).containsInAnyOrder("1->1", "2->2", "3->3");
+      testPipeline.run();
+    } catch (InterruptedException e) {
+      throw new RuntimeException("interrupted.");
+    } finally {
+      p.destroyForcibly();
+    }
+  }
+
+  /** Test TransformProvider. */
+  @AutoService(ExpansionService.ExpansionServiceRegistrar.class)
+  public static class TestTransforms
+      implements ExpansionService.ExpansionServiceRegistrar, Serializable {
+    private final TupleTag<Integer> aTag = new TupleTag<Integer>("aTag") {};
```

Comment:
perhaps label (and name) these "odd" and "even"?

Reply:
Thanks for comment. updated.